src/bit-fields-macros/src/lib.rs

```diff
@@ -0,0 +1,1065 @@
+                    .get_unchecked_mut(i)
+                    .push_str(other.0.get_unchecked(i));
+                i += 1;
+            }
+            while i < other.0.len() {
+                self.0.push(other.0.get_unchecked(i).clone());
+                i += 1;
+            }
+        }
+    }
+    /// Pushes a static string ont `self`.
+    fn push_str(&mut self, s: &str) {
+        self.push(&MultiLineString::from(s));
+    }
+}
+impl From<&str> for MultiLineString {
+    fn from(s: &str) -> Self {
+        Self(s.split('\n').map(String::from).collect())
+    }
+}
+impl From<MultiLineString> for String {
+    fn from(mls: MultiLineString) -> Self {
+        let mut iter = mls.0.into_iter();
+        if let Some(mut base) = iter.next() {
+            for next in iter {
+                base.push('\n');
+                base.push_str(&next);
+            }
+            base
+        } else {
+            String::new()
+        }
+    }
+}
+impl fmt::Display for MultiLineString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for line in &self.0 {
+            writeln!(f, "{}", line)?;
+        }
+        Ok(())
+    }
+}
+
+// TODO Allow writing rustdoc comments on bitfield structs
+/// Convenience macro for errors in [`bitfields`].
+macro_rules! macro_error {
+    ($span:expr,$msg:expr) => {{
+        proc_macro_error::Diagnostic::spanned($span.span(), Level::Error, $msg.into()).emit();
+        return proc_macro::TokenStream::new();
+    }};
+}
+
+/// Procedural macro to generate bit fields.
+///
+/// ```ignore
+/// #[rustfmt::skip]
+/// bit_fields::bitfield!(GeneratedBitField,u32,[
+///     RANGE1, 0..1,
+///     SSE, 2,
+///     SSE1 3,
+///     RANGE2, 4..6,
+///     SSE2, 9,
+///     SSE3, 10,
+///     RANGE3, 12..15,
+///     SSE4, 18
+/// ]);
+/// let bitfield = GeneratedBitField::from(23548);
+/// println!("{}", bitfield);
+/// ```
+/// Prints:
+/// ```test
+/// ┌───────┬────────────┬───┬───────┬───────┬────────────┬───┬───────┬───────┬───┬────────────┬───┬───────┬───┐
+/// │ Bit/s │    00..=00 │ … │    02 │    03 │    04..=05 │ … │    09 │    10 │ … │    12..=14 │ … │    18 │ … │
+/// ├───────┼────────────┼───┼───────┼───────┼────────────┼───┼───────┼───────┼───┼────────────┼───┼───────┼───┤
+/// │ Desc  │     RANGE1 │ … │   SSE │  SSE1 │     RANGE2 │ … │  SSE2 │  SSE3 │ … │     RANGE3 │ … │  SSE4 │ … │
+/// ├───────┼────────────┼───┼───────┼───────┼────────────┼───┼───────┼───────┼───┼────────────┼───┼───────┼───┤
+/// │ Value │          0 │ … │  true │  true │          3 │ … │  true │ false │ … │          5 │ … │ false │ … │
+/// └───────┴────────────┴───┴───────┴───────┴────────────┴───┴───────┴───────┴───┴────────────┴───┴───────┴───┘
+/// ```
+///
+/// # Panics
+///
+/// For a whole load of reason.
+#[allow(clippy::too_many_lines, clippy::expect_used, clippy::unwrap_used)]
+#[proc_macro_error::proc_macro_error]
+#[proc_macro]
+pub fn bitfield(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
```

Comment:
This macro is ~900 lines and it's very hard to read and maintain. There's a lot of stuff going on here apart from parsing and generating the core bitfield structs.
I suggest that this is broken down into multiple components.

What I would really like to see is have one component that does the parsing into an AST.
Then have several components that will traverse the AST and generate the required code. I think you can separate features pretty well if each feature traverses the tree and generates code specifically.

Breaking things down also allows us to write unit tests and implement new things in the future more easily.

Reply:
I have made some large changes, significantly breaking down the code into separate components. I am not sure how to well implement what you suggest here, perhaps the changes I have made are a good enough middle ground?